src/crimson/os/alienstore/alien_store.cc

```diff
@@ -0,0 +1,363 @@
+{
+  logger().debug("{}", __func__);
+  return tp.submit([this, cid] {
+    return store->open_collection(cid);
+  }).then([this] (ObjectStore::CollectionHandle c) {
+    boost::intrusive_ptr<FuturizedCollection> ch (new AlienCollection(c));
+    return seastar::make_ready_future<CollectionRef>(ch);
+  });
+
+}
+
+seastar::future<std::vector<coll_t>> AlienStore::list_collections()
+{
+  logger().debug("{}", __func__);
+
+  std::vector<coll_t> ls;
+  return seastar::do_with(std::move(ls), [=] (auto &ls) {
+    return tp.submit([this, &ls] {
+      return store->list_collections(ls);
+    }).then([&ls] (int) {
+      return seastar::make_ready_future<std::vector<coll_t>>(std::move(ls));
+    });
+  });
+
+}
+
+AlienStore::read_errorator::future<ceph::bufferlist>
+AlienStore::read(CollectionRef ch,
+                 const ghobject_t& oid,
+                 uint64_t offset,
+                 size_t len,
+                 uint32_t op_flags)
+{
+  logger().debug("{}", __func__);
+  return seastar::do_with(ceph::bufferlist{}, [=] (auto &bl) {
+    return tp.submit([=, &bl] {
+      auto c = static_cast<AlienCollection*>(ch.get());
+      return store->read(c->collection, oid, offset, len, bl, op_flags);
+    }).then([&bl] (int r) -> read_errorator::future<ceph::bufferlist> {
+      if (r == -ENOENT) {
+        return crimson::ct_error::enoent::make();
+      } else if (r == -EIO) {
+        return crimson::ct_error::input_output_error::make();
+      } else {
+        return read_errorator::make_ready_future<ceph::bufferlist>(std::move(bl));
+      }
+    });
+  });
+}
+
+AlienStore::get_attr_errorator::future<ceph::bufferptr>
+AlienStore::get_attr(CollectionRef ch,
+                     const ghobject_t& oid,
+                     std::string_view name) const
+{
+  logger().debug("{}", __func__);
+  ceph::bufferptr value;
+  return seastar::do_with(std::move(value), [=] (auto &value) {
+    return tp.submit([=, &value] {
+      auto c =static_cast<AlienCollection*>(ch.get());
+      return store->getattr(c->collection, oid,
+		            static_cast<std::string>(name).c_str(), value);
+    }).then([oid, name, &value] (int r) -> get_attr_errorator::future<ceph::bufferptr> {
+      if (r == -ENOENT) {
+        return crimson::ct_error::enoent::make();
+      } else if (r == -ENODATA) {
+        return crimson::ct_error::enodata::make();
+      } else {
+        return get_attr_errorator::make_ready_future<ceph::bufferptr>(
+          std::move(value));
+      }
+    });
+  });
+}
+
+AlienStore::get_attrs_ertr::future<AlienStore::attrs_t>
+AlienStore::get_attrs(CollectionRef ch,
+                      const ghobject_t& oid)
+{
+  logger().debug("{}", __func__);
+  attrs_t aset;
+  return seastar::do_with(std::move(aset), [=] (auto &aset) {
+    return tp.submit([=, &aset] {
+      auto c = static_cast<AlienCollection*>(ch.get());
+      return store->getattrs(c->collection, oid,
+		             reinterpret_cast<map<string,bufferptr>&>(aset));
```

Comment:
```c++
using attrs_t = std::map<std::string, ceph::bufferptr, std::less<>>;
```

actually i suggest add `ObjectStore` to use map with transparent comparator. so we can avoid this `reinterpret_cast `. but we can leave this for another PR.

Reply:
i am working on this. seems it's non trivial.

Reply:
@liu-chunmei JFYI, my endeavours posted at #32783